larq/models.py

```diff
@@ -1,41 +1,61 @@
 import numpy as np
 from terminaltables import AsciiTable
+import itertools
+from dataclasses import dataclass
 
-__all__ = ["summary"]
-
-
-def sanitize_table(table_data):
-    return [[f"{v:.2f}" if type(v) == float else v for v in row] for row in table_data]
-
+import tensorflow.keras.layers as keras_layers
+import larq.layers as lq_layers
 
-class LayersTable(AsciiTable):
-    def __init__(self, table_data, title=None):
-        super().__init__(sanitize_table(table_data), title=title)
-        self.inner_column_border = False
-        self.justify_columns = {
-            i: "left" if i == 0 else "right" for i in range(len(table_data[0]))
-        }
-        self.inner_footing_row_border = True
-        self.inner_heading_row_border = True
-
-
-class SummaryTable(AsciiTable):
-    def __init__(self, table_data, title=None):
-        super().__init__(sanitize_table(table_data), title=title)
-        self.inner_column_border = False
-        self.inner_heading_row_border = False
+__all__ = ["summary"]
 
+op_count_supported_layer_types = [
+    lq_layers.QuantConv2D,
+    lq_layers.QuantSeparableConv2D,
+    lq_layers.QuantDepthwiseConv2D,
+    lq_layers.QuantDense,
+    keras_layers.Conv2D,
+    keras_layers.SeparableConv2D,
+    keras_layers.DepthwiseConv2D,
+    keras_layers.Dense,
+    keras_layers.Flatten,
+    keras_layers.BatchNormalization,
+    keras_layers.MaxPool2D,
+    keras_layers.AveragePooling2D,
+]
+
+mac_layers = [
+    lq_layers.QuantConv2D,
+    lq_layers.QuantSeparableConv2D,
+    lq_layers.QuantDepthwiseConv2D,
+    lq_layers.QuantDense,
+    keras_layers.Conv2D,
+    keras_layers.SeparableConv2D,
+    keras_layers.DepthwiseConv2D,
+    keras_layers.Dense,
+]
+
+
+def _flatten(lst):
+    return list(itertools.chain.from_iterable(lst))
+
+
+def _bitsize_as_str(bitsize):
+    bitsize_names = {8: "byte", 8 * 1024: "kB"}
 
-def _count_params(weights):
-    """Count the total number of scalars composing the weights.
+    try:
+        return bitsize_names[bitsize]
+    except:
+        raise NotImplementedError()
 
-    # Arguments
-    weights: An iterable containing the weights on which to compute params
 
-    # Returns
-    The total number of scalars composing the weights
-    """
-    return int(sum(np.prod(w.shape.as_list()) for w in weights))
+def _format_table_entry(x, units=1):
+    try:
+        assert not np.isnan(x)
+        if type(x) == str or x == 0 or units == 1:
```

Comment:
would maybe "if type(x) == str: ..." be better because a division is faster than an if statement?

Reply:
They `x==0` is there because otherwise `0` is converted to float and printed as `0.00`.  Similarly, `type(2/1.) == float` while `1 == 1.`, so escaping `units == 1` avoids ints unnecessarily converting to floats.